orangewidget/gui.py

```diff
@@ -1786,7 +1808,19 @@ def do_commit():
     if not checkbox_label:
         btn.setSizePolicy(QSizePolicy.Expanding, QSizePolicy.Maximum)
     checkbox_toggled()
-    setattr(master, commit_name, unconditional_commit)
+
+    decorated = hasattr(commit, "deferred")
+    if decorated:
+        commit.__now = do_commit
+        commit.__deferred = conditional_commit
+    else:
+        commit_name = next(LAMBDA_NAME) if isinstance(commit, LambdaType) else commit.__name__
+        warnings.warn(
+            f"decorate {commit_name} with @gui.deferred and "
+            f"then explicitly call {commit_name}.now or {commit_name}.deferred")
+        setattr(master, 'unconditional_' + commit_name, commit)
+        setattr(master, commit_name, conditional_commit)
```

Comment:
This just triggers warnings when using auto_commit without the decorator. After you add the decorator, it will not complain anymore, but could crash (possibly infrequently and in specific situations not covered by tests) if someone forgets to convert an unconditional_commit call to a .now call.
If this would still (for some transition period) create the unconditional method (with a warning to use .now instead), it would be more bulletproof wrt backwards compatibility albeit uglier.

I am fine with assuming that the decorating will be done carefully and we can leave this as is. I expect searching inside the file for unconditional_commit should be enough and we don't have (m)any cases where that method was being called from another module?

Reply:
It would also raise an exception if somebody forgets to convert deferred calls - see the comment by @VesnaT.

Unconditional commits would not be much of a problem, because they are called after handling inputs and thus tested. Conditional commits are less tested because they often happen as result of use interaction with the interface.

I thought about transitional period (by replacing the original method with deferred one instead of the one that raises exceptions), but I realised it may not make much sense: this transitional period would allow us some time for fixing mistakes that we make while decorating? It's better that incorrect new code fails when written, not half a year later. :)

As a general pattern, this would be as if you wrote new function that will for some time allow somewhat incorrect arguments (but become more strict later) to allow sloppy calls (of this new function!) to be fixed.

Let us rather assume we'll decorate properly. It shouldn't be so hard, it's just about searching for calls of this method.